tests/rules/no-redundant-pipe.test.ts

```diff
@@ -0,0 +1,169 @@
+  valid: [
+    `import { pipe } from "fp-ts/function"
+    pipe(...x);
+  `,
+    `import { pipe } from "fp-ts/function"
+    pipe(value, fn);
+    `,
+    `import { pipe } from "fp-ts/function"
+    pipe(value, pipe(value2, fn));
+    `,
+  ],
+  invalid: [
+    {
+      code: `
+import { pipe } from "fp-ts/function"
+pipe(value);
+`,
+      errors: [
+        {
+          messageId: "redundantPipeWithSingleArg",
+          suggestions: [
+            {
+              messageId: "removePipe",
+              output: `
+import { pipe } from "fp-ts/function"
+value;
+`,
+            },
+          ],
+        },
+      ],
+    },
+    {
+      code: `
+import { pipe } from "fp-ts/function"
+pipe(pipe(pipe(value, fn1), fn2), fn3);
+`,
+      errors: [
+        {
+          messageId: "redundantPipeWithSingleArgInsidePipe",
+          suggestions: [
+            {
+              messageId: "removePipe",
+              output: `
+import { pipe } from "fp-ts/function"
+pipe(pipe(value, fn1), fn2, fn3);
+`,
+            },
+          ],
+        },
+        {
+          messageId: "redundantPipeWithSingleArgInsidePipe",
+          suggestions: [
+            {
+              messageId: "removePipe",
+              output: `
+import { pipe } from "fp-ts/function"
+pipe(pipe(value, fn1, fn2), fn3);
+`,
+            },
+          ],
+        },
+      ],
+    },
+    {
+      code: `
+import { pipe } from "fp-ts/function"
+pipe(pipe(value, fn1), fn2, fn3);
+`,
+      errors: [
+        {
+          messageId: "redundantPipeWithSingleArgInsidePipe",
+          suggestions: [
+            {
+              messageId: "removePipe",
+              output: `
+import { pipe } from "fp-ts/function"
+pipe(value, fn1, fn2, fn3);
+`,
+            },
+          ],
+        },
+      ],
+    },
+    {
+      code: `
+import { pipe } from "fp-ts/function"
+pipe(
+  // foo
+  pipe(value, fn1),
+  fn2
+);`,
+      errors: [
+        {
+          messageId: "redundantPipeWithSingleArgInsidePipe",
+          suggestions: [
+            {
+              messageId: "removePipe",
+              output: `
+import { pipe } from "fp-ts/function"
+pipe(
+  // foo
+  value, fn1,
+  fn2
+);`,
+            },
+          ],
+        },
+      ],
+    },
+    {
+      code: `
+import { pipe } from "fp-ts/function"
+pipe(
+  value,
+);
+`,
+      errors: [
+        {
+          messageId: "redundantPipeWithSingleArg",
+          suggestions: [
+            {
+              messageId: "removePipe",
+              output: `
+import { pipe } from "fp-ts/function"
+value;
+`,
+            },
+          ],
+        },
+      ],
+    },
+    {
+      code: `
+import { pipe } from "fp-ts/function"
+pipe(
+  // foo
+  value,
+);
+`,
+      errors: [
+        {
+          messageId: "redundantPipeWithSingleArg",
+          suggestions: [
+            {
+              messageId: "removePipe",
+              // TODO: ideally we would preserve the comment here but I'm not sure how
```

Comment:
If I'm reading this correctly, this is due to the fact that we're synthesizing a new AST node. It's hard to preserve trivias (comments, whitespaces, etc) when manipulating the AST because it's a lossy representation of the source code.
This is why I tend to lean towards reading AST but writing tokens (aka characters). It's harder to work with, but it's less invasive for the final user.

Do you think this fix could be implemented without generating an AST? Accidentally removing comments is a show-stopper in my opinion, I think it's a fix that can produce invalid syntax in some edge cases is still preferable to a silent removal of a comment

Reply:
Thanks for bringing this up! I agree it's not good if we accidentally remove a comment. To be clear, this is only an issue for comments inside the `pipe`—comments outside are preserved.


> Do you think this fix could be implemented without generating a new AST node?

Yes, I think it would be the same as the fix we used to have for `no-redundant-flow` before https://github.com/buildo/eslint-plugin-fp-ts/pull/224/commits/b76044afa46f006869624450daacc46bf2034ed5 which was to fix https://github.com/buildo/eslint-plugin-fp-ts/issues/222. Maybe there is another way to fix https://github.com/buildo/eslint-plugin-fp-ts/issues/222?

That said, in my opinion it's extremely error prone to manipulate code as text rather than as an AST—although this unfortunately seems to be the status quo for ESLint rules, as far as I can tell. As far as I understand, there is no information about comments in TSESTree's representation of the AST. However, there is some information about comments in TypeScript's AST. I wonder if we could implement the fix using TypeScript's AST instead. Do you have any experience doing this inside of ESLint rules?

Reply:
I don't have direct experience with Typescript AST, no, but if it supports not blowing away the comments maybe it's worth a try. 

Otherwise we can think of a more ad-hoc fix for the specific use case

Reply:
I think the ad-hoc fix would be to revert https://github.com/buildo/eslint-plugin-fp-ts/commit/b76044afa46f006869624450daacc46bf2034ed5 and then find another way to workaround https://github.com/buildo/eslint-plugin-fp-ts/issues/222, but I'm not sure how else we can workaround this?